File: Data Structures/LinkedList/DoublyLL/C/DoublyLL.c
#include<stdio.h>
#include<stdlib.h>
#include<conio.h>

struct node
{
    int data;
    struct node *next;
    struct node *prev;
};

struct node *head;

struct node *createNode(int value)
{
    struct node *newnode;
    newnode = (struct node*)malloc(sizeof(struct node));
    newnode -> data = value;
    newnode -> next = NULL;
    newnode -> prev = NULL;
    return newnode;
}

void insertATBeg(int value)
{
    struct node *newnode = createNode(value);
    if(head==NULL)
    {
        newnode -> next = head;
        head = newnode;
    }
    else
    {
        newnode -> next = head;
        head -> prev = newnode;
        head = newnode;
    }
}

void insertAtEnd(int value)
{
    struct node *newnode=createNode(value);
    struct node *temp;
    temp=head;
    while (temp->next!=NULL)
    {
        //traversing till last element
        temp=temp->next;
    }
    if (temp->next==NULL)
    {
        temp->next=newnode;
        newnode->prev=temp;
    }
    else
    printf("Something went wrong\n");   
}

void deletionAtBeg(){
    if (head!=NULL)
    {
        struct node *temp;
        temp=head;
        head=head->next;
        head->prev = NULL;
        free(temp);
        printf("Deletion at begining successfull \n");
    }
    else
    {
        printf("Nothing to delete \n");
    }   
}

void deletionAtEnd(){
    struct node* temp=head;
    while(temp->next->next!=NULL)
    {
        temp=temp->next;
    }
    free(temp->next->next);
    temp->next=NULL;
    printf("Deletion at the end is successfull");    
}

void show()
{
    struct node * temp;
    temp=head;
    while (temp!=NULL)
    {
        //traversing till last element's next part
        printf("%d   ",temp->data);
        temp=temp->next;
    }
    printf("");
    
}

int count()
{
    struct node *temp;
    temp=head;
    int c=0;
    while(temp!=NULL){
        c++;
        temp=temp->next;
    }
    return c;
}

void insertAtBet(int value,int pos)
{
    struct node *newnode=createNode(value);
    struct node *temp = head;
    int c=2;
    while(temp!=NULL){
        c++;
        temp=temp->next;

        if (c==pos)
        {
            newnode->next = temp->next;
            temp->next = newnode;
            break;
        }   
    }
}

void deleteAtBet(int pos){
    struct node *temp = head;
    struct node *temp2 = head->next;
    int c=2;
    while(temp!=NULL){
        c++;
        temp=temp->next;
        temp2=temp2->next;

        if (c==pos)
        {
            temp->next=temp2->next;
            free(temp2);
            break;
        }   
    }

}

void menu()
{
    printf("1. Insertion from beginning \n");
    printf("2. Insertion for end \n");
    printf("3. Deletion from beginning \n");
    printf("4. Deletion from end \n");
    printf("5. Size of List \n");
    printf("6. Insertion in between \n");
    printf("7. Deletion in between \n"); 
    printf("8. Show my list \n");
    printf("9. Exit() \n");
    
}

int main()
{
    int c,x,pos;
    while (1)
    {

        menu();
        printf("Enter your choice : ");
        scanf("%d",&c);

        switch (c)
        {

        case 1:
            /* inserting at begining in the list */
            printf("Enter the value : ");
            scanf("%d",&x);
            /* Creating a list */
            insertATBeg(x);
            
            break;
        
        case 2:
            /* inserting at end */
            printf("Enter the value : ");
            scanf("%d",&x);
            /* Creating a list */
            insertAtEnd(x);
            break;
        
        case 3:
            /* deletion from beg */
            deletionAtBeg();
            break;
        
        case 4:
            /* Deletion at end */
            deletionAtEnd();
            break;
        
        case 5:
            /* size of list */
            printf("Number of Element : %d\n",count());
            break;
        
        case 6:
            /* insertion at a position */
            printf("Enter the postion : ");
            scanf("%d",&pos);
            printf("Enter the value : ");
            scanf("%d",&x);
            /* Creating a list */
            insertAtBet(x,pos);

            break;
        
        case 7:
            /* deletion ata position */
            printf("Enter the pos : ");
            scanf("%d",&pos);
            /* Creating a list */
            deleteAtBet(pos);
            break;
        
        case 8:
            /* Showing the list */
            show();
            break;
        
        case 9:
            /* exit */
            exit(0);
            break;
        
        default:
            break;
        }
    }
}

File: Data Structures/LinkedList/DoublyLL/python/doublyLL.py
class node:
    def __init__(self,value):
        self.data=value
        self.next=None
        self.prev=None

class DoubleLinkedList:
    def __init__(self):
        self.head=None

    def insertAtBeg(self,value):
        newnode = node(value)
        if self.head==None:
            self.head=newnode
        else:
            self.head.prev = newnode
            newnode.next=self.head
            self.head=newnode

    def insertAtEnd(self,value):
        newnode=node(value)
        if self.head == None:
            self.head=newnode
        else:
            temp=self.head
            while temp.next is not None:
                temp=temp.next
            temp.next=newnode
            newnode.prev=temp

    def insertAtBet(self,value,pos):
        pass
    
    def deleteAtBeg(self):
        temp = self.head
        self.head=self.head.next
        
    def deleteAtEnd(self):
        temp=self.head
        while temp.next.next is not None:
            temp=temp.next
        temp.next = None
        
    def deleteAtBet(self,pos):
        pass

    def show(self):
        temp = self.head
        while(temp is not None):
            print(temp.data)
            temp=temp.next

    def revshow(self):
        temp = self.head
        while(temp.next is not None):
            temp=temp.next
        while(temp is not None):
            print(temp.data)
            temp=temp.prev
    
def menu():
        print('------------------------------------------------------------------------------')
        print()
        print('1. Insertion at beginning')
        print('2. Insertion at end')
        print('3. Insertion at between')
        print('4. Deletion at beginning')
        print('5. Deletion at end')
        print('6. Deletion at between')
        print('7. Show')
        print('8. Reverse show')
        print('9. Exit()')
        print()
        print('------------------------------------------------------------------------------')
        



if __name__ == '__main__':
    ll = DoubleLinkedList()
        
    while(True):

        menu()
        ch = int(input('Enter your choice : '))
        if ch==1:
            value= int(input('ENter your data : '))
            ll.insertAtBeg(value)
        elif ch==2:
            value= int(input('ENter your data : '))
            ll.insertAtEnd(value)
        elif ch==3:
            value= int(input('ENter your data : '))
            pos= int(input('ENter your position : '))
            ll.insertAtBet(value,pos)
        elif ch==4:
            ll.deleteAtBeg()
        elif ch==5:
            ll.deleteAtEnd()
        elif ch==6:
            pos=int('Enter the position : ')
            ll.deleteAtBet(pos)
        elif ch==7:
            print('***************************************************************************')
            ll.show()
            print('***************************************************************************')
        elif ch==8:
            print('***************************************************************************')
            ll.revshow()
            print('***************************************************************************')
        elif ch==9:
            exit()
        else:
            print("Enter some valid option")

    

